src/cc/BPFTable.h

```diff
@@ -68,11 +68,104 @@ class BPFTableBase {
   size_t capacity_;
 };
 
+template <class ValueType>
+class BPFArrayTable : protected BPFTableBase<int, ValueType> {
+private:
+  // this class is used to allow updating a table element using the
+  // table[index] = value syntax
+  class BPFArrayTableWriter {
+    public:
+      BPFArrayTableWriter(BPFArrayTable<ValueType>& t, const int& i):
+      table(t), index(i) {}
+      BPFArrayTableWriter& operator=(ValueType const& rhs) {
+        table.update_value(index, rhs);
+        return *this;
+      }
+      operator ValueType const() {
+        return table.get_value(index);
+      }
+    private:
+      BPFArrayTable<ValueType>& table;
+      const int& index;
+  };
+
+public:
+  BPFArrayTable(BPFModule* bpf_module, const std::string& name)
+      : BPFTableBase<int, ValueType>(bpf_module, name) {
+    int type = bpf_module->table_type(name);
+    if (type != BPF_MAP_TYPE_ARRAY &&
+        type != BPF_MAP_TYPE_PERCPU_ARRAY)
+      throw std::invalid_argument("Table '" + name + "' is not an array table");
+  }
+
+  ValueType get_value(const int& index) {
+    ValueType res;
+    if (!this->lookup(const_cast<int*>(&index), &res))
+      throw std::invalid_argument("index out of range");
+    return res;
+  }
+
+  // updates an element
+  void update_value(const int& index, const ValueType& value) {
+    if (!this->update(const_cast<int*>(&index), const_cast<ValueType*>(&value)))
+      throw std::invalid_argument("Error updating element");
```

Comment:
why is it impossible to return multiple values?
I'd really like to see existing throw in get_value() go away as well.
Dropping operator [] is also good thing, since it's semantics are ambiguous.
I especially don't like BPFHashTableWriter operator[].
I see no need to make bpf maps looks like std::maps. They are different and that should be obvious from api.